File: apps/api/src/schemas/voice.py
"""Pydantic models for voice command processing.

Defines lightweight NLP intent payloads and responses for the voice
command endpoint. Keeps parsing simple and rule-based.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class VoiceAction(str, Enum):
    """Supported voice command actions."""

    NAVIGATE = "navigate"
    STOP = "stop"
    EMERGENCY_STOP = "emergency_stop"
    UNLOCK_CONTAINER = "unlock_container"
    TAKE_PHOTO = "take_photo"
    STATUS = "status"
    RETURN_HOME = "return_home"
    CREATE_DELIVERY = "create_delivery"


class VoiceCommandRequest(BaseModel):
    """Incoming voice text to parse/execute."""

    text: str = Field(..., min_length=1, max_length=256, description="Transcript text")
    user_role: str = Field(
        "operator",
        pattern=r"^(operator|admin|recipient)$",
        description="Role of the speaker for permission checks",
    )
    robot_id: str = Field("robot-001", description="Target robot for the command")
    confirm: bool = Field(
        False, description="Whether the user already confirmed a sensitive action"
    )


class VoiceCommandIntent(BaseModel):
    """Parsed intent from raw voice text."""

    action: VoiceAction
    target_floor: Optional[int] = None
    recipient_name: Optional[str] = None
    tag_id: Optional[str] = None
    camera_source: Optional[str] = None
    requires_confirmation: bool = False
    confidence: float = Field(0.5, ge=0.0, le=1.0)


class SupportedCommand(BaseModel):
    """A command that the voice system can handle."""

    phrase: str
    action: VoiceAction
    description: str
    example: str
    role_required: str = "operator"


class VoiceCommandResponse(BaseModel):
    """Response from the voice endpoint with parsed intent and execution info."""

    success: bool
    text: str
    robot_id: str
    intent: VoiceCommandIntent
    message: str
    executed: bool = False
    eta_seconds: Optional[int] = None
    timestamp: datetime


class SupportedCommandsResponse(BaseModel):
    """List of all supported voice commands."""

    commands: list[SupportedCommand]

File: apps/web/lib/api/voice.ts
/**
 * Voice command API client for Stair-Doc.
 */

import { OpenAPI } from "@/lib/api/client";

export type VoiceAction =
  | "navigate"
  | "stop"
  | "emergency_stop"
  | "unlock_container"
  | "take_photo"
  | "status"
  | "return_home"
  | "create_delivery";

export interface VoiceCommandIntent {
  action: VoiceAction;
  target_floor?: number | null;
  tag_id?: string | null;
  camera_source?: string | null;
  recipient_name?: string | null;
  requires_confirmation: boolean;
  confidence: number;
}

export interface SupportedCommand {
  phrase: string;
  action: VoiceAction;
  description: string;
  example: string;
  role_required: string;
}

export interface SupportedCommandsResponse {
  commands: SupportedCommand[];
}

export interface VoiceCommandResponse {
  success: boolean;
  text: string;
  robot_id: string;
  intent: VoiceCommandIntent;
  message: string;
  executed: boolean;
  eta_seconds?: number | null;
  timestamp: string;
}

export interface VoiceCommandPayload {
  text: string;
  user_role: "admin" | "operator" | "recipient";
  robot_id?: string;
  confirm?: boolean;
}

interface ApiEnvelope<T> {
  success: boolean;
  data: T;
  message: string;
  timestamp: string;
}

export class VoiceApiError extends Error {
  constructor(
    message: string,
    public statusCode: number,
    public body?: unknown,
  ) {
    super(message);
    this.name = "VoiceApiError";
  }
}

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const url = `${OpenAPI.BASE}${path}`;

  const response = await fetch(url, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...init?.headers,
    },
  });

  if (!response.ok) {
    const body = await response.json().catch(() => null);
    throw new VoiceApiError(
      body?.detail ?? body?.message ?? `Request failed: ${response.status}`,
      response.status,
      body,
    );
  }

  const envelope: ApiEnvelope<T> = await response.json();

  if (!envelope.success) {
    throw new VoiceApiError(envelope.message, response.status, envelope);
  }

  return envelope.data;
}

export async function sendVoiceCommand(
  payload: VoiceCommandPayload,
): Promise<VoiceCommandResponse> {
  return apiFetch<VoiceCommandResponse>("/api/v1/voice/command", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function fetchSupportedCommands(): Promise<SupportedCommand[]> {
  const result = await apiFetch<SupportedCommandsResponse>("/api/v1/voice/commands");
  return result.commands;
}
